# <center> <h1>Assignment 6</h1> </center>
<center> <h1>EIN 4933/6935 Python for Data Science Summer 2020</h1> </center>

You will build a decision tree model using a data set related to breast cancer (Breast Cancer Wisconsin (Diagnostic) Data Set). You can have more info about the data set in this <a href="http://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+%28diagnostic%29">web-site</a>. 

You can use this link: https://sites.google.com/site/yasinunlu/home/research/new1/breast_cancer.csv to dowload the data in csv format. In this dataset there are 30 features and one response variable. It is highly likely to predict breast cancer by using a decision tree model and given the set of predictors for a number of observations.

The column named "diagnosis" shows whether a patient is diagnosed as "malignant" (M) or "benign" (B). Use this dataset to answer the questions below. You can create as many cells as you desire for all solutions.<br/>

## Questions:
1) Read the data into a dataframe from csv file source. Download link is given above. <br/>
2) Remove any irrelevant data columns from the dataframe. If there are any missing values in the dataset, they should be replaced through a data imputation method. Any missing rows should be deleted from the dataset. Each column data type must be converted into an appropriate data type. The data types can be obtained either in data info files or can be determined by inspection.<br/>
3) Convert categorical variables in "diagnosis" column into integer numerical values.<br/>
4) Transform the data so that it fits within a specific scale (0 and 1).<br/>
5) Create two dataframes: response and features. Split these dataframes into train and test parts.<br/>
6) Train a decision tree model using train part of the data.<br/>
7) Generate predictions over the test data.<br/>
8) Calculate the accuracy. <br/>
a) Write code that calculates the accuracy over test data.<br/>
b) Call the built-in function from scikit learn package to calculate the accuracy over test data.<br/>
9) Report confusion matrix.<br/>
10) Report feature importance results. Create a dataframe that has two columns: one for the name of the feature and one for the associated score.<br/>

In [118]:
import pandas as pd
df = pd.read_csv("https://sites.google.com/site/yasinunlu/home/research/new1/breast_cancer.csv")
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

#### Step-1 Problem Definition:

In [130]:
df.shape

(569, 32)

The dataset has 569 rows and 32 columns. Evidently we have two possible outcomes as response variable: M for malignant and B for benign.

In [126]:
df.diagnosis.drop_duplicates()

0     M
19    B
Name: diagnosis, dtype: object

Our analysis will lead to understand which model(s) are the most helpful in predicting cancer type (i.e malignant or benign). We will perform feature importance analysis. We will also try to improve the model results.

#### Step-2: Data Cleaning and Preprocessing:

Since the column 'id' cannot be used as a predictor, we will remove it from the current dataframe.

In [127]:
df_clean = df.drop(axis=1, columns='id')
df_clean.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

Let's take a look at if the dataset has any nulls or NaNs.

In [134]:
print(df_clean.isnull().sum())

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64


In [135]:
print(df_clean.isna().sum())

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64


Since there is no nulls/NaNs in the dataset, we will not perform any imputations. 

#### Step-3 (optional): Feature Extraction/Addition:

In this step, we will convert any columns with categorical values to numerical values. 

In [137]:
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
df_clean['diagnosis'] = labelencoder.fit_transform(df['diagnosis'])

In [138]:
df_clean.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1        17.99         10.38          122.80     1001.0   
1          1        20.57         17.77          132.90     1326.0   
2          1        19.69         21.25          130.00     1203.0   
3          1        11.42         20.38           77.58      386.1   
4          1        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

In [140]:
df_clean.diagnosis.drop_duplicates()

0     1
19    0
Name: diagnosis, dtype: int32

Values of 1 for M and 0 for B are populated in diagnosis column.

#### Step-4: Data Scaling (optional):
Since diagnosis column can only take 0 or 1 and features columns have wide varirty of values, we will normalize the features in a range of 0 and 1.

In [155]:
from sklearn.preprocessing import StandardScaler
sc = MinMaxScaler()
my_matrix = sc.fit_transform(df_clean)
df_clean = pd.DataFrame(my_matrix, columns = df_clean.columns)
df_clean.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0        1.0     0.521037      0.022658        0.545989   0.363733   
1        1.0     0.643144      0.272574        0.615783   0.501591   
2        1.0     0.601496      0.390260        0.595743   0.449417   
3        1.0     0.210090      0.360839        0.233501   0.102906   
4        1.0     0.629893      0.156578        0.630986   0.489290   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0         0.593753          0.792037        0.703140             0.731113   
1         0.289880          0.181768        0.203608             0.348757   
2         0.514309          0.431017        0.462512             0.635686   
3         0.811321          0.811361        0.565604             0.522863   
4         0.430351          0.347893        0.463918             0.518390   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0       0.686364  ...      0.620776       0.141525         0.668310   
1       0.379798  ...      0.606901       0.303571         0.539818   
2       0.509596  ...      0.556386       0.360075         0.508442   
3       0.776263  ...      0.248310       0.385928         0.241347   
4       0.378283  ...      0.519744       0.123934         0.506948   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    0.450698          0.601136           0.619292         0.568610   
1    0.435214          0.347553           0.154563         0.192971   
2    0.374508          0.483590           0.385375         0.359744   
3    0.094008          0.915472           0.814012         0.548642   
4    0.341575          0.437364           0.172415         0.319489   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              0.912027        0.598462                 0.418864  
1              0.639175        0.233590                 0.222878  
2              0.835052        0.403706                 0.213433  
3              0.884880        1.000000                 0.773711  
4              0.558419        0.157500                 0.142595  

[5 rows x 31 columns]

In [156]:
df_clean.diagnosis.drop_duplicates()

0     1.0
19    0.0
Name: diagnosis, dtype: float64

We need to convert the data type of diagnosis back to int type.

In [162]:
df_clean.diagnosis = df_clean.diagnosis.astype(int)

In [163]:
df_clean.diagnosis.drop_duplicates()

0     1
19    0
Name: diagnosis, dtype: int32

In [167]:
df_clean.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1     0.521037      0.022658        0.545989   0.363733   
1          1     0.643144      0.272574        0.615783   0.501591   
2          1     0.601496      0.390260        0.595743   0.449417   
3          1     0.210090      0.360839        0.233501   0.102906   
4          1     0.629893      0.156578        0.630986   0.489290   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0         0.593753          0.792037        0.703140             0.731113   
1         0.289880          0.181768        0.203608             0.348757   
2         0.514309          0.431017        0.462512             0.635686   
3         0.811321          0.811361        0.565604             0.522863   
4         0.430351          0.347893        0.463918             0.518390   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0       0.686364  ...      0.620776       0.141525         0.668310   
1       0.379798  ...      0.606901       0.303571         0.539818   
2       0.509596  ...      0.556386       0.360075         0.508442   
3       0.776263  ...      0.248310       0.385928         0.241347   
4       0.378283  ...      0.519744       0.123934         0.506948   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    0.450698          0.601136           0.619292         0.568610   
1    0.435214          0.347553           0.154563         0.192971   
2    0.374508          0.483590           0.385375         0.359744   
3    0.094008          0.915472           0.814012         0.548642   
4    0.341575          0.437364           0.172415         0.319489   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              0.912027        0.598462                 0.418864  
1              0.639175        0.233590                 0.222878  
2              0.835052        0.403706                 0.213433  
3              0.884880        1.000000                 0.773711  
4              0.558419        0.157500                 0.142595  

[5 rows x 31 columns]

#### Step-5: Data Splitting:
Let's start with creating two dataframes: response and features. We will split these dataframe into train (80% of all rows) and test (20% of all rows) parts.

In [170]:
features = df_clean.iloc[:,1:]
response = df_clean[['diagnosis']]

In [171]:
features.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0     0.521037      0.022658        0.545989   0.363733         0.593753   
1     0.643144      0.272574        0.615783   0.501591         0.289880   
2     0.601496      0.390260        0.595743   0.449417         0.514309   
3     0.210090      0.360839        0.233501   0.102906         0.811321   
4     0.629893      0.156578        0.630986   0.489290         0.430351   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0          0.792037        0.703140             0.731113       0.686364   
1          0.181768        0.203608             0.348757       0.379798   
2          0.431017        0.462512             0.635686       0.509596   
3          0.811361        0.565604             0.522863       0.776263   
4          0.347893        0.463918             0.518390       0.378283   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                0.605518  ...      0.620776       0.141525         0.668310   
1                0.141323  ...      0.606901       0.303571         0.539818   
2                0.211247  ...      0.556386       0.360075         0.508442   
3                1.000000  ...      0.248310       0.385928         0.241347   
4                0.186816  ...      0.519744       0.123934         0.506948   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0    0.450698          0.601136           0.619292         0.568610   
1    0.435214          0.347553           0.154563         0.192971   
2    0.374508          0.483590           0.385375         0.359744   
3    0.094008          0.915472           0.814012         0.548642   
4    0.341575          0.437364           0.172415         0.319489   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0              0.912027        0.598462                 0.418864  
1              0.639175        0.233590                 0.222878  
2              0.835052        0.403706                 0.213433  
3              0.884880        1.000000                 0.773711  
4              0.558419        0.157500                 0.142595  

[5 rows x 30 columns]

In [172]:
response.head()

diagnosis
0          1
1          1
2          1
3          1
4          1

In [173]:
from sklearn.model_selection import train_test_split
my_result_list = train_test_split(features, response, test_size=0.2, random_state=0)
features_train, features_test, response_train, response_test = my_result_list

In [174]:
print("Number of total rows in complete features dataframe: ", len(features))
print("Number of total rows in features_train dataframe: ", len(features_train))
print("Number of total rows in features_test dataframe: ", len(features_test))
print("Number of total rows in complete response dataframe: ", len(response))
print("Number of total rows in response_train dataframe: ", len(response_train))
print("Number of total rows in response_test dataframe: ", len(response_test))

Number of total rows in complete features dataframe:  569
Number of total rows in features_train dataframe:  455
Number of total rows in features_test dataframe:  114
Number of total rows in complete response dataframe:  569
Number of total rows in response_train dataframe:  455
Number of total rows in response_test dataframe:  114


#### Step-6: Model Selection, Model Fitting and Model Evaluation:

We will train Decision Tree model using the data that we prepared:<br/>

In [177]:
#Decision Tree model fitting
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(random_state = 0)
classifier.fit(features_train, response_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [179]:
response_pred = classifier.predict(features_test)
print('Predictions: ', response_pred)
print('Actuals: ', list(response_test['diagnosis']))

Predictions:  [1 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 1 1 1 1 1 0 0 1 0 0 1 0 1 0 1 0 1 0 1 0
 1 0 1 0 0 1 0 1 1 0 0 0 1 1 1 1 0 0 0 1 0 1 1 1 1 0 0 1 1 1 1 1 0 0 1 0 0
 1 0 0 0 0 0 1 1 1 0 1 0 0 0 1 1 0 1 0 1 0 0 1 0 0 0 0 0 0 1 1 0 1 0 1 0 0
 1 1 0]
Actuals:  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0]


Manual calculation to estimate the accuracy.

In [186]:
my_array = response_pred==list(response_test['diagnosis'])
accuracy = my_array.sum()/len(response_pred)
print('Accuracy Score on test data: ', accuracy)

Accuracy Score on test data:  0.9122807017543859


Use accuracy_score() built-in function to estimate the accuracy.

In [180]:
from sklearn.metrics import accuracy_score
print('Accuracy Score on test data: ', accuracy_score(response_test, response_pred))

Accuracy Score on test data:  0.9122807017543859


#### Step-7: Report Feature Importance:

Feature importance

In [195]:
importance = classifier.feature_importances_
feature_names = features.columns.to_list()

# summarize feature importance
for i, score in enumerate(importance):
    print('%s: %.5f' % (feature_names[i], score))

radius_mean: 0.00000
texture_mean: 0.00976
perimeter_mean: 0.02255
area_mean: 0.00000
smoothness_mean: 0.03894
compactness_mean: 0.00000
concavity_mean: 0.00292
concave points_mean: 0.00000
symmetry_mean: 0.01648
fractal_dimension_mean: 0.00000
radius_se: 0.00000
texture_se: 0.00000
perimeter_se: 0.00000
area_se: 0.01216
smoothness_se: 0.00148
compactness_se: 0.00000
concavity_se: 0.00000
concave points_se: 0.01648
symmetry_se: 0.00000
fractal_dimension_se: 0.00761
radius_worst: 0.00000
texture_worst: 0.00947
perimeter_worst: 0.01491
area_worst: 0.13591
smoothness_worst: 0.00000
compactness_worst: 0.00000
concavity_worst: 0.00930
concave points_worst: 0.69283
symmetry_worst: 0.00000
fractal_dimension_worst: 0.00919
